[1단계 - 로또 구현] 줄리(한승아) 미션 제출합니다.

jbilee opened the pull request:
안녕하세요 오리!
이번에 랜디와 페어 프로그래밍을 진행한 줄리입니다 :)

로또를 구현하면서 고민했던 부분을 여쭙고 싶었는데 마감 시간 맞추기에 벅차 미션만 제출하게 되어 아쉽기만 합니다. 우선 셀프체크 링크와 함께 PR 올립니다:
https://prolog.techcourse.co.kr/studylogs/3833

코드 리뷰해주셔서 감사합니다!

KJungW committed:
docs(readme): readme 작성

작성 내용
- 프로젝트 소개
- 요구사항
- 전체 실행 과정

KJungW committed:
test(lotto): 로또 관련 테스트 작성

테스트 내용
- 로또 객체를 생성할 수 있다
- 로또번호를 정렬해서 저장한다
- 중복되는 번호가 있는지 검증한다
- 입력된 번호들이 1부터 45사이인지 검증한다
- 입력된 번호가 6개인지 검증한다

jbilee committed:
feat(lotto): 로또 도메인 구현

작업 내용
- 작성된 테스트에 맞춰 로또 구현
- 로또에 사용하기 위해 일부 검증 로직 추가

KJungW committed:
feat(input): 로또 구매 금액 입력

구현 내용
- 사용자로부터 로또 구매 금액 입력받음
- 관련 검증 로직 추가

jbilee committed:
feat(input): 당첨 기준 입력

작업 내용
- 지난 주 당첨 번호 입력
- 보너스 볼 입력
- 관련 검증 로직 추가

KJungW committed:
fix(input): InputValidator 로직 수정

작업 내용
- null 체크 추가

KJungW committed:
refactor(lotto): 로또 가격 상수 생성

작업 내용
- 로또 클래스에 로또 가격 상수 추가
- 기존의 로또 가격 대체

KJungW committed:
fix(input): 당청 번호 검증 로직 수정

작업 내용
- 스트림의 map() 작동하도록 map() 결과값 처리

KJungW committed:
test(validator): 입력값 검증 테스트

테스트 내용
- 입력값이 null인 경우 예외 발생
- 입력값이 빈 값인 경우 예외 발생
- 입력값이 숫자가 아닐 경우 예외 발생
- 입력된 구매 금액이 1000단위가 아닌 경우 예외 발생
- 당첨 번호가 숫자가 아닌 경우 예외 발생
- 보너스 번호가 입력된 당첨 번호와 중복되면 예외 발생

jbilee committed:
test(validator): 비즈니스 로직 검증 테스트

작업 내용
- 리스트 내에 중복되는 번호가 있으면 예외 발생
- 리스트 내 숫자가 범위를 벗어난 경우 예외 발생
- 단일 숫자가 범위를 벗어난 경우 예외 발생
- 리스트 크기가 잘못된 경우 예외 발생

KJungW committed:
feat(lotto_service): 금액에 따른 로또 구매

작업 내용
- 구입 금액으로 로또 개수 계산

jbilee committed:
test(lotto_service): 금액에 따른 로또 구매 테스트

작업 내용
- 구매 금액에 해당하는 로또 개수를 계산할 수 있다

jbilee committed:
feat(lotto_service): 로또 생성

작업 내용
- 정해진 갯수만큼 로또를 생성
- 랜덤 번호 생성기 구현

KJungW committed:
feat(lotto_service): 당첨 통계 계산

작업 내용
- 발행된 로또의 당첨 등수 계산
- 당첨 등수를 열거형으로 관리

jbilee committed:
feat(application): 로또 애플리케이션 구성

작업 내용
- 애플리케이션의 메인 메소드 생성
- 로또 컨트롤러 생성을 위한 설정 클래스 생성
- 로또 컨트롤러 생성

KJungW committed:
fix(lotto_service): 당첨 통계 계산 수정

작업 내용
- 당첨 통계가 제대로 수행되지 않는 점 수정

jbilee committed:
feat(lotto_service): 수익률 계산

작업 내용
- 총 상금과 구매 금액을 토대로 수익률 계산

KJungW committed:
feat(output): 구매한 로또 출력

작업 내용
- 구매한 로또 개수 출력
- 발행된 로또 출력

KJungW committed:
refactor(input): 프롬프트 추가

작업 내용
- 사용자 입력을 안내하는 프롬프트 추가

KJungW committed:
refactor(output): 출력 형식 수정

작업 내용
- 출력 포맷에 빈 줄 추가

jbilee committed:
feat(output): 당첨 결과 출력

작업 내용
- 당첨 통계와 총 수익률 출력

jbilee committed:
refactor(lotto_service): 미사용 메서드 삭제

작업 내용
- 로또 단일 발행 메소드 삭제

jbilee committed:
test(lotto_service): 로또 서비스 테스트

작업 내용
- 구매한 개수만큼의 로또를 생성할 수 있다

KJungW committed:
fix(lotto_service): 등수 계산 버그 수정

작업 내용
- 2등이 필터링되지 않는 버그 수정

KJungW committed:
test(lotto_service): 등수 계산 테스트

작업 내용
- 발행된 로또의 당첨 등수를 찾을 수 있다.

jbilee committed:
test(lotto_service): 로또 서비스 테스트

작업 내용
- 수익률을 올바르게 계산할 수 있다

KJungW committed:
refactor(application): 전체적인 리펙토링

작업 내용
- 메서드 길이 15줄 이하로 수정
- 메서드 뎁스 1단계로 제한

KJungW committed:
refactor(application): 패키지명 오타 수정

작업 내용
- constant 패키지 이름 오타 수정

KJungW committed:
style(test): 스타일 정리

KJungW committed:
refactor(input): 입력 검증 리펙토링

작업 내용
- 미사용 상수 삭제
- 구입 금액이 0일 경우에 대한 검증 추가

KJungW committed:
test(input): 테스트케이스 추가

추가 내용
-  구매 금액이 0보다 작거나 같을 경우 예외 발생 처리

KJungW committed:
style(test): 스타일 정리

jbilee committed:
style(application): 스타일 정리

jbilee committed:
refactor(application): 예외 메세지 열거형 리팩토링

작업 내용
- ExceptionMessage 내 미사용 상수 삭제

jbilee committed:
docs(readme): 리드미 문서 수정

작업 내용
- 미션 repository명이 잘못 기재되어 있던 부분 수정

jbilee committed:
fix(application): 패키지 오타 수정

작업 내용
- validator 패키지명 오타 수정

jinyoungchoi95 commented on src/test/java/lotto/service/LottoServiceTest.java:
```diff
@@ -0,0 +1,63 @@
+
+        List<WinningTier> actualTiers = lottoService.findWinningTiers(lottos, winningLotto);
+
+        assertThat(actualTiers).containsExactlyInAnyOrderElementsOf(Arrays.stream(WinningTier.values()).toList());
```
> AssertJ를 활용하여 특정 조건을 검증하는 테스트 코드를 작성할 수 있는가?
> AssertJ의 .isStrictlyBetween(), .containsExactlyInAnyOrderElementsOf() 등 상세한 메소드를 활용해 특정 조건을 검증하도록 테스트를 구현할 수 있습니다.

`isStrictlyBetween`는 어떤 상황에서 사용하고 현재 미션에서 어떻게 활용할 수 있나요? 
`containsExactlyInAnyOrderElementsOf`는 어떻게 기능동작을 하나요?

jbilee replied:
`isStrictlyBetween` 메소드는 Integer나 LocalDate처럼 정량적으로 비교가 가능한 값이 명시한 범위 내에 있는지를 확인하는 메소드인데, 생각해보니 테스트에서 로또 번호의 범위(1~45 사이)를 검증하는 데에 사용할 수 있을 것 같습니다.
```java
@Test
void 로또_번호의_범위를_검증한다() {
    assertThat(1).isStrictlyBetween(0, 46);
}
```
랜덤값을 생성하는 함수가 1~45 사이의 숫자를 생성해줬는지 확인하는 테스트를 작성해서 활용해볼 생각입니다.

`containsExactlyInAnyOrderElementsOf`는 두 리스트를 비교해 정확히 같은 요소들을 가지고 있는지를 판단하는데, 이때 요소들의 순서는 고려하지 않습니다. `List.of(1, 2, 3)`와 `List.of(3, 1, 2)`를 비교하면 true하지만, `List.of(1, 2, 3)`와 `List.of(1, 2, 2)`를 비교하면 한 리스트에는 3이 없고 2가 대신 들어있기 때문에 false합니다.

코드 중 `containsExactlyInAnyOrderElementsOf`가 쓰이는 줄에 코멘트를 달아주셨기에 제가 알맞게 활용한게 아닌건지 생각해봤는데 따로 떠오르는 건 없는 것 같습니다.. 🥹 일단 다른 코드를 리팩토링하면서 계속 생각해보려는데 이런 경우에는 어떻게 나아가는게 더 현명할까요?

jinyoungchoi95 replied:
문제가 있어서 코멘트를 드렸던 것은 아니고 학습로그 작성하실 때 설명이 부족해보여 추가로 질문드렸던 것이었어요.

이후에도 학습로그 작성하실때는 지금처럼 구체적으로 작성해주시면 리뷰어도 같이 확인하면서 가지고 있는 기반지식을 쌓아나갈 수 있는 좋은 기회가 될 것 같아요.

jinyoungchoi95 commented on src/test/java/lotto/domain/LottoTest.java:
```diff
@@ -0,0 +1,57 @@
+package lotto.domain;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatIllegalArgumentException;
```
> JUnit5와 AssertJ의 주요 차이점은 무엇인가?
> JUnit5과 AssertJ의 주요 차이점은 assertion 메소드의 활용 방식으로, AssertJ의 경우 메소드 체이닝을 통해 좀 더 가독성 높은 테스트 코드를 작성할 수 있다는 장점이 있습니다.

메소드 체이닝을 사용하면 어떻게 가독성이 더 올라가나요?

jbilee replied:
저는 아래 이유로 가독성이 올라간다고 생각합니다:
- 메소드 체이닝 없이 한 함수로만 구현하면 assert하려는 상황과 예상되는 결과값 등을 함께 매개변수로 전달하게 되는데, 이때 여러개의 매개변수들을 한눈에 파악하기 어려울 수 있음, 메소드 체이닝으로 나누면 분리점이 확실해서 가독성 측면에서 도움이 됨
- 문법적으로도 더 자연스러운 문장 형식이 만들어져서 직관적이고 빠르게 읽힘
(예: `assertNotEquals()` vs `assertThat().isNotEqualTo()`)

jinyoungchoi95 commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,34 @@
+
+public class Lotto {
+
+    public static final int MIN_LOTTO_NUMBER = 1;
+    public static final int MAX_LOTTO_NUMBER = 45;
+    public static final int LOTTO_NUMBER_COUNT = 6;
+    public static final int LOTTO_PRICE = 1000;
```
각 상수는 `Lotto`라는 도메인이 가지고 있는 고유한 정보이기 때문에 private으로 외부로 노출하지 않고 본인이 도메인지식을 컨트롤할 수 있도록 변경해보세요.

외부에서 활용해야한다면 `Lotto`라는 도메인을 잘 설계했는지 고민해보는 포인트가 될 것 같네요.

jbilee replied:
코드를 다시 살펴보니 외부에서 사용자의 입력값 검증을 대신 해주도록 구현을 한 것이 `Lotto`가 고유로 갖고 있어야 할 정보가 노출된 원인인 것 같습니다. 이 부분은 사용자로부터 입력을 받는 과정을 일부 수정하면서 `Lotto` 도메인의 정보를 보호하는 방향을 고안해 개선해보려 합니다!

jinyoungchoi95 commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,34 @@
+    private final List<Integer> numbers;
+
+    public Lotto(List<Integer> numbers) {
+        LogicValidator.validateDuplication(numbers);
+        LogicValidator.validateRange(numbers, MIN_LOTTO_NUMBER, MAX_LOTTO_NUMBER);
+        LogicValidator.validateSize(numbers, LOTTO_NUMBER_COUNT);
```
`LogicValidator`가 필요한 이유가 무엇인가요? 도메인이 본인이 처리하면 안되나요?

jbilee replied:
검증 메소드로 인해 도메인의 볼륨이 필요 이상으로 커질 것 같아서 비즈니스 로직에 사용되는 값들만 검증하기 위한 전용 validator를 만들었는데, 이로 인해 `Lotto` 도메인의 정보들이 노출되는 이슈가 가장 큰 부작용인 것 같습니다. 게다가 코드를 다시 보니 뷰에서 받아온 입력값에 대한 검증에 `InputValidator`와 `LogicValidator`를 둘 다 쓰고 있는 것도 확인했습니다 😂 처음에 생각했던 `LogicValidator`의 필요성이 타당한지 다시 생각해보려 합니다.

jinyoungchoi95 commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,34 @@
+    }
+
+    public List<Integer> getNumbers() {
+        return Collections.unmodifiableList(this.numbers);
```
<img width="602" alt="스크린샷 2025-02-13 오후 8 41 48" src="https://github.com/user-attachments/assets/0e595453-d89f-4b33-beab-cf75138712d4" />

인텔리제이에에서 왜 까맣게 표현하고 있을까요?

jbilee replied:
`Collections.unmodifiableList()`에 대한 설명에서 아래 문장을 발견했습니다.
```
This method may return its argument if the argument is already unmodifiable.
```
`Lotto` 생성자에서 `toList()` 메소드를 통해 이미 불변성이 보장되는 리스트를 `this.numbers`에 저장하기 때문에 `Collections.unmodifiableList()`는 받은 매개변수를 그대로 반환하게 되고, 인텔리제이가 이를 redundant하다고 체크해준다는 점 확인했습니다..! 👀

jinyoungchoi95 replied:
인텔리제이가 항상 정답은 아니지만 기본 설정에서 표기하고 있는 위험 문구들은 거의 수정이 필요한 경우이므로 코드짜실때 항상 체크해두시면 좋을 것 같아요.

jinyoungchoi95 commented on src/test/java/lotto/domain/LottoTest.java:
```diff
@@ -0,0 +1,57 @@
+        List<Integer> numbers = List.of(1, 2, 3, 4, 5, 6);
+        Lotto lotto = new Lotto(numbers);
+
+        assertThat(lotto.getNumbers()).containsExactlyElementsOf(numbers);
```
테스트하고자하는 대상은 정상 생성인데 `getNumbers`라는 메소드에 의존하고 있네요. `getNumbers`가 main 코드에서 안쓰인다면 테스트를 위한 메소드가 되는 것 아닐까요? `getNumbers` 를 사용하지 않고 테스트할 수 있는 방법은 없을까요?

jbilee replied:
로또 객체 인스턴스가 제대로 생성되는지에 대한 테스트인데 말씀하신대로 인스턴스의 getter로 상태 값만 검증하는 것은 적합하지 않은 것 같습니다. "제대로 생성되었다"를 판단하는 기준을 객체 생성 시 예외가 발생하는지 아닌지로 변경해보려고 합니다. 도메인 내부적으로 검증을 거쳐 정상 생성 여부를 확인하는 것이 더 확실한 도메인 설계와 테스트일 것 같다는 생각이 듭니다.

jinyoungchoi95 replied:
`equals`를 재정의하여 테스트하는 방법도 있습니다. `isEqualTo`가 `equals`를 재정의하였을 때와 하지 않았을 때의 차이를 확인해보시면 이해가실거에요.

jinyoungchoi95 commented on src/test/java/lotto/domain/LottoTest.java:
```diff
@@ -0,0 +1,57 @@
+                .withMessage(ExceptionMessage.DUPLICATED_NUMBERS.getContent());
+    }
+
+    @DisplayName("입력된 번호들이 1부터 45사이인지 검증한다.")
```
threshold는 0과 46이므로 46만을 검증할 것이 아니라 추가로 0이 있는 경우도 테스트케이스에 넣어보아야하지 않을까요?

jbilee replied:
확인했습니다! 지금 보면 너무 당연한 미스인데 당시에는 안보였던 것 같습니다.. 😅 꼼꼼히 체크해주셔서 감사합니다!

jinyoungchoi95 commented on src/test/java/lotto/domain/LottoTest.java:
```diff
@@ -0,0 +1,57 @@
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+
+class LottoTest {
```
`hasNumber`와 `findMatches`는 테스트할 필요가 없는걸까요?

jbilee replied:
인텔리제이 테스트 커버리지에는 `hasNumber`와 `findMatches`도 커버가 되었다고 표시되어서 충분한 테스트가 되었다고 여겼는데, 이는 다른 클래스의 로직을 테스트하는 단계에서 사용되었을 뿐이지 `Lotto` 클래스 자체의 단위 테스트에 해당하지 않는 것 같습니다. (실제로 어느 테스트로 인해 해당 메소드들이 커버리지에 포함됐는지 다시 찾을 때 살짝 헤맸습니다..) `Lotto` 클래스 내에서의 쓰임새를 직접적으로 테스트해봐야 할 것 같습니다.

jinyoungchoi95 commented on src/main/java/lotto/constant/WinningTier.java:
```diff
@@ -0,0 +1,37 @@
+import java.util.List;
+import lotto.domain.WinningCondition;
+
+public enum WinningTier {
```
`WinningCondition`이 한번 더 래핑되어야하는 이유가 있을까요?

```java
.filter(tier -> tier.getCondition().isWinningCondition(matchedCount, isBonusNumberMatched))
```
에서 메시지를 전달하는 형태도 아니라 getter로 꺼내오고 있는데 객체로써의 역할을 하고 있는지 고민해보시면 좋을 것 같아요.

jbilee replied:
`WinningCondition` 클래스는 당첨 여부를 판단하는 로직을 수행하기 위해, `WinningTier` enum은 당첨 랭크에 따른 상수를 저장하기 위해 생성했으나 `getCondition`으로 인해 결국엔 불필요한 중간 단계가 추가된 것 같습니다. `WinningTier`가 직접 당첨 조건을 확인하는 방향으로 개선해보려 합니다!

jinyoungchoi95 commented on src/main/java/lotto/config/ApplicationConfiguration.java:
```diff
@@ -0,0 +1,30 @@
+import lotto.view.InputView;
+import lotto.view.OutputView;
+
+public class ApplicationConfiguration {
```
config 설정이 필요한 서비스인가요..? 사용하는 측에서 단순하게 구현하면 될 것 같은데 Configuration이 들어가서 복잡하게 객체 생성을 할 필요가 없어보이네요.

jbilee replied:
컨트롤러가 활용해야 할 서비스와 뷰를 코드 흐름을 파악하기 쉽도록 한곳에서부터 전달할 목적으로 config를 사용했는데, 결국 중간에서 getter로 전달만 하도록 설계되어 불필요한 레이어가 한층 더 생겨버린 것 같습니다. 명확한 목적이 없는 중간 매개체 사용은 지양하는 방향으로 수정하려 합니다.

jinyoungchoi95 commented on src/main/java/lotto/constant/ExceptionMessage.java:
```diff
@@ -0,0 +1,20 @@
+package lotto.constant;
+
+public enum ExceptionMessage {
```
"공통화"된 객체는 최대한 지양하고 도메인 객체를 잘 만드는데 집중해보세요.

jbilee replied:
공통화된 객체라고 하면 자바 애플리케이션에 상시 적재(탑재)되어 여러 곳에서 쓰이는 객체라는 뜻인 것 같은데 제가 이해한게 맞는지 여쭙고 싶습니다..!

jinyoungchoi95 commented on src/main/java/lotto/service/LottoService.java:
```diff
@@ -0,0 +1,44 @@
+import lotto.domain.WinningLotto;
+import lotto.utility.RandomGenerator;
+
+public class LottoService {
```
`layered architecture`를 활용하신 것 같은데, 1레벨 미션에서 spring과 관련된 모든것은 잊으시고 도메인 객체를 어떻게 잘 만들지를 고민해보세요.

`Service`라는 의미없는 객체에 두기보다는 이율을 계산하거나 랭크를 계산하는 `domain` 패키지에 적절하게 들어갈만한 도메인 객체 설계를 배워나가는 것이 1레벨의 최종 목표입니다.

jbilee replied:
오리의 코멘트를 읽고 도메인과 서비스의 정의에 대해 다시 생각해봤습니다. (그 사이 오늘차 백엔드 강의에서도 도메인과 서비스에 대한 코치와 동료 크루들의 의견도 들어볼 수 있었습니다..!)

이율 계산이나 랭크 계산은 도메인이 관리해야 하는 정보와 직접적으로 연관되는데 이걸 전부 `Service`가 외부의 입장에서 처리하도록 억지로 맞춘 것 같다는 느낌이 듭니다. 도메인의 역할에 더 부합하는 설계를 고민해보겠습니다.

jinyoungchoi95 commented on src/main/java/lotto/domain/WinningCondition.java:
```diff
@@ -0,0 +1,32 @@
+package lotto.domain;
+
+public class WinningCondition {
```
전반적으로 단위 테스트가 부족해요. 어디서부터 어디까지 테스트짜야할지 감이 오지 않는다면 일단 모든 테스트를 다 짜보는 경험을 해보고 거기서부터 필요없다고 생각하는 테스트들을 줄여가보면 줄리만의 테스트 짜는 기준을 만들어나갈 수 있을 것 같아요.

jbilee replied:
피드백 감사합니다! 말씀대로 테스트를 추가로 작성하면서 단위 테스트의 의미도 다시 생각해보려고 합니다!

jinyoungchoi95 reviewed (changes requested):
안녕하세요 줄리 😄

로또 미션의 리뷰어를 맡은 오리입니다.

첫 미션 잘 구현해주셨네요. 다음 단계 넘어가기 전 같이 이야기하면 좋을 내용 코멘트 남겨두었으니 확인부탁드려요.

궁금하거나 고민이 되는 부분이 있으시다면 언제든 pr 코멘트 또는 dm으로 요청 부탁드립니다.
감사합니다 🙇‍♂️

jbilee committed:
refactor(lotto): 로또 클래스 리팩토링

작업 내용
- 불필요한 불변성 컬렉션 래퍼 삭제

jbilee committed:
refactor(validation): 검증 로직 리팩토링

작업 내용
- `LogicValidator`가 아닌 로또 도메인에서 로또 관련 입력값을 검증하도록 수정
- `LogicValidator` 삭제

jbilee committed:
test(validation): 테스트 삭제

작업 내용
- `LogicValidator` 삭제로 인한 관련 테스트 삭제

jbilee committed:
refactor(lotto): 로또 도메인 검증 로직 리팩토링

작업 내용
- 불필요하게 전달되는 매개변수를 없애고 로또 도메인의 클래스 변수를 직접적으로 사용하도록 수정

jbilee committed:
refactor(lotto): 당첨 등수 enum 리팩토링

작업 내용
- `WinningCondition` 대신 `WinningTier` enum 클래스에서 당청 등수를 직접 확인하도록 수정
- 미사용 클래스(`WinningCondition`) 삭제

jbilee committed:
test(utility): 랜덤값 생성 기능 테스트 추가

테스트 내용
- 입력한 최대 범위까지의 난수 리스트를 생성한다

jbilee committed:
test(lotto): 로또 도메인 테스트 수정

작업 내용
- `Lotto` 객체 정상 생성에 대한 검증을 예외 발생 여부로 확인하도록 변경

jbilee committed:
test(lotto): 로또 도메인 테스트 수정

작업 내용
- 로또 번호 범위 테스트에 1 미만의 값으로 테스트 조건 추가

jbilee committed:
test(lotto): 로또 도메인 테스트 추가

테스트 내용
- 특정 숫자가 로또 번호 리스트에 존재하는지 확인한다
- 로또 번호와 일치하는 숫자의 개수를 구한다

jbilee committed:
refactor(output): 출력 서비스 삭제

작업 내용
- 출력 뷰를 제어하는 `OutputService`를 삭제하고, 해당 서비스의 로직을 뷰로 이동

jbilee committed:
refactor(input): 입력 서비스 삭제

작업 내용
- 입력 뷰를 제어하는 `InputService`를 삭제하고, 해당 서비스의 로직을 뷰로 이동

jbilee committed:
refactor(output): 출력 뷰 리팩토링

작업 내용
- 사용되지 않는 메소드(`printLine`) 삭제

jbilee committed:
refactor(view): 뷰 관련 코드 리팩토링

작업 내용
- 뷰와 관련된 공통 상수를 삭제하고 뷰 내부에서 메소드 호출 시에만 관련 문자열을 출력하도록 수정

jbilee committed:
refactor(view): 입력 뷰 리팩토링

작업 내용
- 입력 뷰 생성 시 Scanner를 매개로 받지 않고 자체적으로 생성하도록 수정

jbilee committed:
refactor(application): 애플리케이션 config 삭제

작업 내용
- `ApplicationConfiguration`에서 컨트롤러에 뷰, 로또 서비스를 주입하던 부분을 `Application`에서 진행하도록 수정
- `ApplicationConfiguration` 클래스 삭제

jbilee committed:
refactor(lotto_service): 로또 서비스 리팩토링

작업 내용
- `LottoService`에서 처리하던 이율 계산 로직을 별도 도메인에서 처리하도록 수정
- 이율 계산 전용 도메인 생성 (`Vendor`)

jbilee committed:
refactor(lotto_service): 로또 서비스 리팩토링

작업 내용
- `LottoService`에서 처리하던 당첨 등수 계산 로직을 별도 도메인에서 처리하도록 수정
- 이율 계산 전용 도메인 생성 (`Scoreboard`)

jbilee committed:
refactor(lotto_service): 로또 서비스 리팩토링

작업 내용
- `LottoService`의 로또 발행 및 로또 개수 계산 로직을 `Vendor` 클래스가 처리하도록 수정
- `LottoService` 클래스 삭제

jbilee committed:
test(vendor): Vendor 도메인 테스트 작성

테스트 내용
- `LottoServiceTest` 내 로또 발행 및 이율 계산 관련 테스트를 `VendorTest`로 이동

jbilee committed:
test(scoreboard): Scoreboard 도메인 테스트 작성

테스트 내용
- `LottoServiceTest` 내 당첨 등수 계산 로직에 대한 테스트를 `ScoreboardTest`로 이동
- `LottoServiceTest` 테스트 삭제

jbilee committed:
refactor(lotto): 로또 도메인 로직 리팩토링

작업 내용
- 보너스 번호의 유효성 여부를 `WinningLotto` 클래스에서 검증하도록 수정

jbilee committed:
refactor(global): 공통 상수 삭제

작업 내용
- 다수의 클래스에서 쓰이는 `ExceptionMessage` 열거형 삭제

jbilee committed:
test(lotto): 로또 도메인 테스트 수정

작업 내용
- 테스트에서 특정 오류 메세지 문자열을 확인하는 부분 삭제

jbilee committed:
test(validation): 입력 검증 테스트 수정

작업 내용
- 테스트에서 특정 오류 메세지 문자열을 확인하는 부분 삭제

jbilee committed:
fix(tier): 등수 열거형 관련 로직 오류 수정

작업 내용
- `WinningTier` 열거형이 당첨 등수를 제대로 반환하지 않던 이슈 수정

jbilee committed:
test(lotto): 당첨 로또 테스트 작성

테스트 내용
- 구매한 로또의 당첨 등수를 찾을 수 있다

jbilee committed:
refactor(validation): 구매 금액 검증 로직 리팩토링

작업 내용
- 구매 금액을 `Vendor` 클래스에서 검증하도록 수정

jbilee committed:
test(validation): 구매 금액 검증 테스트 수정

작업 내용
- `InputValidatorTest`의 구매 금액 검증 테스트를 `VendorTest`로 이동

jbilee committed:
refactor(validation): 입력 검증 로직 리팩토링

작업 내용
- 당첨 번호 입력값에 대한 검증에서 검증 결과를 불필요하게 변수로 저장하던 부분 수정

jbilee committed:
refactor(lotto): 로또 도메인 리팩토링

작업 내용
- 로또 도메인의 클래스 변수를 public에서 private으로 변경
- `Vendor` 클래스에 최대 난수값에 대한 클래스 변수 추가

jbilee committed:
test(lotto): 당첨 로또 관련 테스트 추가

테스트 내용
- 보너스 번호가 당첨 번호와 중복될 경우 예외가 발생한다

jbilee committed:
style(global): 컨벤션에 맞춰 코드 스타일 수정

jbilee committed:
test(lotto): 당첨 로또 도메인 테스트 추가

테스트 내용
- 보너스 번호가 1과 45 사이가 아닐 경우 예외가 발생한다

jinyoungchoi95 commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,53 @@
+        this.numbers = numbers.stream().sorted().toList();
+    }
+
+    private void validateDuplication(List<Integer> numbers) {
```
`Set`으로 받으면 중복을 검증할 필요가 없지 않을까요?

jbilee replied:
처음부터 Set으로 받는 시나리오는 생각치 못했습니다! 아이러니하게도 전달 받은 List를 Set으로 새로 저장해 둘의 크기를 비교하는 복잡하고 비효율적인 로직을 짜버린 것 같습니다. 코드를 다시 살펴보면서 `RandomGenerator`에 매직넘버를 쓴 부분도 캐치해 함께 수정하려고 합니다.

추가: `Lotto` 클래스 로직을 수정하다가 깨달았는데, 사용자가 직접 입력한 당첨 번호 중에 중복된 번호가 있는지를 검증하기 위해 `Lotto` 도메인에 해당 메소드를 추가하게 됐었습니다. 외부에서 봤을 땐 사용자 입력값에 대한 검증으로 여겨질 것 같아서`validateDuplication` 메소드를 `InputValidator`로 이동할지에 대해서도 고민해봤는데, 6개의 중복되지 않는 로또 번호가 필요하다는 조건은 `Lotto` 도메인만의 정보라고 느껴져 현재로서는 이동시키지 않는 것이 적합한 것 같습니다.. 이 부분은 도메인 설계에 대해 학습한 내용을 복기하면서 계속 고민해보겠습니다!

jinyoungchoi95 commented on src/test/java/lotto/domain/LottoTest.java:
```diff
@@ -0,0 +1,92 @@
+                Arguments.of(new Lotto(List.of(6, 8, 9, 10, 11, 12)), 1)
+        );
+    }
+}
```
github에서 왜 빨간색으로 문제를 표기하고 있을까요?

jbilee replied:
해당 문제는 파일의 끝지점(프로그램 실행 시 오류가 발생하지 않도록 끝지점을 정확히 명시하도록 함)을 알려주는 개행문자가 없어서 GitHub이 표시해주는 것인데, 작업할 때 개행 설정을 제대로 하지 않아 휴먼에러가 발생한 것 같습니다. OS에 따라 개행 포맷이 다르다는 것도 이번에 처음 알았습니다! (윈도우: CRLF 또는 LF, 리눅스/맥: LF) 2단계 진행하면서 함께 수정하겠습니다.

jinyoungchoi95 commented on src/test/java/lotto/domain/LottoTest.java:
```diff
@@ -0,0 +1,92 @@
+        List<Integer> numbers = List.of(1, 2, 3, 4, 5, 6);
+        Lotto lotto = new Lotto(numbers);
+
+        assertThat(lotto.hasNumber(4)).isEqualTo(true);
+        assertThat(lotto.hasNumber(7)).isEqualTo(false);
```
테스트의 단위라는 것을 바라볼때 어떤 상황을 피드백 받을 것이냐가 중요한 분기가 되는데요. 현재 테스트는
- 번호가 리스트에 없으면 false를 반환한다
- 번호가 리스트에 있으면 true를 반환한다

로 각각 테스트가 원하는 피드백이 달라요. 물론 하나의 메소드를 사용한 것이기 때문에 똑같은 메소드 호출로 만든 기능이기는 하지만 돌아오는 피드백 자체가 다르기 때문에 두 케이스는 분리하여 테스트를 작성하는 것이 좋습니다.

jbilee replied:
단순히 “리스트 내에 존재하는지를 확인한다”라는 너무 큰 개념으로만 생각해서 true한 케이스와 false한 케이스는 엄연히 다른 상황이라는 생각을 못했던 것 같습니다. 더욱 명확한 테스트를 작성할 수 있도록 유념하겠습니다!

jinyoungchoi95 commented on src/main/java/lotto/domain/Vendor.java:
```diff
@@ -0,0 +1,50 @@
+    private final int purchaseAmount;
+
+    public Vendor(int purchaseAmount) {
+        this.validatePurchaseAmount(purchaseAmount);
```
본인 메소드호출에 this를 붙이는 기준이 있을까요?

jbilee replied:
저는 개인적으로 프로그래밍을 할 때 도움이 돼서 this를 붙여주고 있습니다. 메소드 앞에 this가 있으면 같은 클래스 내에 구현되어 있을테니 해당 메소드를 어디에서 찾아봐야 하는지 바로 알 수 있고, 인스턴스 메서드로 구현했던 것을 나중에 static으로 바꾸려고 하면 IntelliJ가 캐치해주기도 해서 실수를 좀 더 줄이기 위해 활용하고 있습니다. this가 추가되면 static 메소드를 호출했을 때와 시각적으로 차이가 있어서 수많은 코드를 계속 수정해나가야 하는 상황에서 길라잡이 역할을 해준다고 느낍니다.

검색해보니 컴파일러가 알아서 처리해주기 때문에 this를 붙이는 것은 redundant한 것이고, 따라서 지양하는 것이 좋다는 의견이 있는데 저는 이 부분은 잘 공감이 되지는 않는 것 같습니다.. 컴파일러는 컴파일러고, 직접 수작업을 해야 하는 개발자와는 다르기 때문에 그렇게 느끼는 것 같습니다. 혹시 오리는 redundant하다고 생각되시는지 궁금합니다 😯

jinyoungchoi95 commented on src/main/java/lotto/domain/Vendor.java:
```diff
@@ -0,0 +1,50 @@
+
+    public List<Lotto> issueLottos() {
+        int lottoCount = this.calculateLottoCount();
+        RandomGenerator randomGenerator = new RandomGenerator();
```
로또를 살때마다 `RandomGenerator` 객체가 불필요하게 생성될 수 있겠네요.

jbilee replied:
로또 구매가 현 요구사항에서는 프로그램의 라이프사이클 중 한 번 일어나서 `issueLottos`에서 생성하도록 두었는데 비즈니스 로직의 변동 가능성을 고려하면 메소드 밖에서 생성하는 것이 알맞은 것 같습니다--다만 그렇다면 외부 어디에서 생성해야 효율적일지 고민이 됩니다.. Vendor의 인스턴스 변수로 두면 Vendor가 존재하는 이상 사용되지 않아도 함께 존재하게 돼서, 불필요한 인스턴스 변수를 추가하는 대신 `issueLottos`를 호출할 때 주입시켜주는 방식으로 수정해보려 합니다.

추가: 위 내용에 대해 찾아보니 객체 생성이 자주 일어나면 오버헤드가 발생할 수 있고, 이는 상시로 멤버 변수를 두는 것보다 비용이 더 클 수 있다는 것을 알게 됐습니다. `issueLottos`를 호출할 때 주입시키는 방식 역시 해당 메소드가 호출되는 회차마다 객체 생성 비용이 발생하기 때문에 Vendor의 멤버 변수에 저장하는 것이 더 나은 선택인 것 같습니다.

jinyoungchoi95 commented on src/main/java/lotto/constant/WinningTier.java:
```diff
@@ -0,0 +1,49 @@
+import java.util.List;
+
+public enum WinningTier {
+    FIRST(6, false, 2000000000),
+    SECOND(5, true, 3000000),
+    THIRD(5, false, 1500000),
+    FOURTH(4, false, 50000),
+    FIFTH(3, false, 5000),
```
```suggestion
    FIRST(6, false, 2_000_000_000),
    SECOND(5, true, 3_000_000),
    THIRD(5, false, 1_500_000),
    FOURTH(4, false, 50_000),
    FIFTH(3, false, 5_000),

```
숫자의 가독성을 위해 _를 사용할 수 있습니다.

jbilee replied:
생활꿀팁 감사합니다! 😆

jinyoungchoi95 commented on src/main/java/lotto/constant/WinningTier.java:
```diff
@@ -0,0 +1,49 @@
+    }
+
+    public static WinningTier find(int matches, boolean hasBonusMatch) {
+        List<WinningTier> allTiers = List.of(WinningTier.values());
+        return allTiers.stream()
```
```suggestion
        return Arrays.stream(WinningTier.values())
```
로 바로 사용할 수도 있겠네요.

jbilee replied:
자료구조와 알고리즘만큼 어려운게 Stream 사용법인 것 같습니다.... 피드백 감사합니다!!

jinyoungchoi95 commented on src/test/java/lotto/domain/ScoreboardTest.java:
```diff
@@ -0,0 +1,31 @@
+    @Test
+    void 발행된_로또의_당첨_등수를_찾을_수_있다() {
+        List<Lotto> lottos = List.of(
+                new Lotto(List.of(1, 2, 3, 4, 5, 6)),
```
`new Lotto`에 `List.of`를 매번 테스트코드에 작성하기 귀찮아 보이지않으신가요?
```java
public Lotto(int... numbers) {
```
와 같은 생성자가 있으면 편할 것 같아보이네요.

jbilee replied:
프리코스 때 `int[]` 같은 일반 배열 대신 더 다양한 API를 지원하는 컬렉션을 지향한다는 피드백을 받아서 그동안 의무적으로 List를 고수해왔던 것 같습니다. 말씀하신대로 생성자에서 `int...`를 받도록 변경해보려고 했는데, 중간에 검증 메소드들이 다 매개변수로 List를 받도록 짜여져 있어서 이 부분을 어떻게 처리하면 효율적인 코드가 될지 고민이 됩니다.. 확실히 `new Lotto`를 생성할 때마다 List를 전달하는 것이 부담되어서, 우선 `Lotto`의 생성자를 overload하는 방식으로 접근해보려고 합니다.

jinyoungchoi95 reviewed (approved):
안녕하세요 줄리 😄

피드백 잘 반영해주셨네요. 몇가지 코멘트 남겨둔 것은 다음 단계에 함께 반영부탁드려요.
미션 코멘트 알림을 받기 힘들어 궁금하신점은 다음 단계 pr에 함께 올려주시면 다음 단계에서 같이 확인하겠습니다.

궁금하거나 고민이 되는 부분이 있으시다면 언제든 pr 코멘트 또는 dm으로 요청 부탁드립니다.
감사합니다 🙇‍♂️
